DevTools/UserspaceEmulator/Emulator.cpp

```diff
@@ -1523,4 +1529,26 @@ int Emulator::virt$clock_nanosleep(FlatPtr params_addr)
     return rc;
 }
 
+int Emulator::virt$readlink(FlatPtr params_addr)
+{
+    Syscall::SC_readlink_params params;
+    mmu().copy_from_vm(&params, params_addr, sizeof(params));
+
+    if (params.path.length > PATH_MAX) {
+        return -ENAMETOOLONG;
+    }
+    auto path = mmu().copy_buffer_from_vm((FlatPtr)params.path.characters, params.path.length);
+    char host_buffer[PATH_MAX] = {};
```

Comment:
Please, please, don't use `PATH_MAX`, except for checking syscall inputs against it (like it's done a few lines above — that one is OK). In particular, don't ever use it for any buffer sizes.

Using a buffer of size `PATH_MAX` here will both overallocate for the common case — most symlink paths are in fact very short, e.g. we have `/bin/sh` → `Shell` (5 bytes) and `/bin/[` → `test` (4 bytes) — and not work in case the symlink path is actually longer than `PATH_MAX`.

I have previously made `Core::File::read_link()` work properly for both short links (the common case) and arbitrarily long links, see https://github.com/SerenityOS/serenity/commit/47d83800e1c27c0a26d3ee432973d49a3f83f299 and https://github.com/SerenityOS/serenity/commit/d89843f96fa7673f409bb8c33e043e2355b257f1. We should make it work as well when UE-ed.

Reply:
Regarding the LibC implementation: How is this better? Now in the case of overlong pathnames, the result is truncated. Taken from your linked commit:

```
__RETURN_WITH_ERRNO(rc, min((size_t)rc, size), -1);
```
That looks like a problem, because it can sometimes return something that is not an error code *and* not the target of the link.

I agree that over-allocation isn't particularly nice, but that's how `virt$realpath` does it, and it seems comparable to a `malloc`. So I went with the same thing, for consistency.

If we get rid of `PATH_MAX`, then we run into the issue that we need to allocate `params.buffer.size` amount of memory, without knowing the actual size yet that will be required. Since that variable is controlled by the program being emulated, this could be a malicious value like 2GiB. Where do you draw the line? `PATH_MAX` seemed like a good idea. We could also validate that the entire `params.buffer` region is mapped and writable, but that still leaves the question open how to behave if it doesn't work.

Please tell me (or implement) how to do that in a better way.

Also, have you experimented with extremely long paths? I run into trouble at a total path length of 2K already (`PATH_MAX = 4K`), so something fishy is going on anyways.

Reply:
> Regarding the LibC implementation: How is this better? Now in the case of overlong pathnames, the result is silently truncated.

Yes, this is what POSIX seems to want:

> If the buf argument is not large enough to contain the link content, the first bufsize bytes shall be placed in buf. <...> Upon successful completion, these functions shall return the count of bytes placed in the buffer. Otherwise, these functions shall return a value of -1, leave the buffer unchanged, and set errno to indicate the error.

(So clearly they consider the case of a too-short buffer a "successful completion", not an error.)

The `File::read_link()` function, on the other hand, can be made to handle this properly, which I've done.

> I agree that over-allocation isn't particularly nice, but that's how `virt$realpath` does it

That one, too, needs fixing :slightly_smiling_face:

> Please tell me (or implement) how to do that in a better way.

I'd just use `params.buffer.size` here.

> If we get rid of PATH_MAX, then we run into the issue that we need to allocate params.buffer.size amount of memory, without knowing the actual size yet that will be required. Since that variable is controlled by the program being emulated, this could be a malicious value like 2GiB. Where do you draw the line?

I mean, this emulated program is not being sandboxed by us (UE) in any way, in fact we're running _in its process_. In the same way that an in-process library does not sandbox a potentially malicious program using it. If it asks for 2 GB, so be it. It's somebody else's job to impose memory limits on us IMO.

> Also, have you experimented with extremely long paths?

Not really. It'd be a good idea to do that and fix up all the bugs we encounter.

Reply:
Let's move this to the "Discussions" tab that @awesomekling enabled recently: https://github.com/SerenityOS/serenity/discussions/4357